dll/cpl/sysdm/general.c

```diff
@@ -598,73 +606,75 @@ static VOID GetSystemVersion(HWND hwnd)
 
     QueryPerformanceCounter(&Counter);
     QueryPerformanceFrequency(&Frequency);
 
-    return Counter.QuadPart / Frequency.QuadPart;
+    return (Counter.QuadPart * 1000) / Frequency.QuadPart;
 }
 
-ULONGLONG GetSeconds(VOID)
+static VOID GetSystemUptime(HWND hwndDlg)
 {
-    ULONGLONG (WINAPI * pGetTickCount64)(VOID);
-    ULONGLONG Ticks64;
-    HMODULE hModule = GetModuleHandleW(L"kernel32.dll");
+    HWND hUptimeLabel;
+    WCHAR szBuf[64];
+    ULONG cSeconds;
 
-    pGetTickCount64 = (PVOID)GetProcAddress(hModule, "GetTickCount64");
     if (pGetTickCount64)
     {
-        return pGetTickCount64() / 1000;
+        RosUptime = pGetTickCount64();
     }
-
-    hModule = LoadLibraryW(L"kernel32_vista.dll");
-
-    if (!hModule)
+    else
     {
-        return GetSecondsQPC();
+        RosUptime = GetTickCountQPC();
     }
 
-    pGetTickCount64 = (PVOID)GetProcAddress(hModule, "GetTickCount64");
-
-    if (pGetTickCount64)
-    {
-        Ticks64 = pGetTickCount64() / 1000;
-    }
-    else
+    hUptimeLabel = GetDlgItem(hwndDlg, IDC_UPTIME);
+    if (!hUptimeLabel)
     {
-        Ticks64 = GetSecondsQPC();
+        return;
     }
+    
+    cSeconds = RosUptime / 1000;
+    StringCchPrintfW(szBuf, _countof(szBuf), szUptimeFormat,
+                     cSeconds / (60*60*24),
+                     (cSeconds / (60*60)) % 24,
+                     (cSeconds / 60) % 60,
+                     cSeconds % 60);
 
-    FreeLibrary(hModule);
-    return Ticks64;
+    SetWindowTextW(hUptimeLabel, szBuf);
 }
 
-VOID GetSystemUptime(HWND hwnd)
+static VOID InitSystemUptime(HWND hwndDlg)
 {
-    HWND hRosUptime;
-    WCHAR szBuf[64], szStr[64];
-    ULONG cSeconds;
-
-    hRosUptime = GetDlgItem(hwnd, IDC_UPTIME);
-    if (!hRosUptime)
+    /* Load time format string */
+    if (LoadStringW(hApplet, IDS_UPTIME_FORMAT, szUptimeFormat, _countof(szUptimeFormat)) == 0)
     {
         return;
     }
-    if (!LoadStringW(hApplet, IDS_UPTIME_FORMAT, szStr, _countof(szStr)))
+
+    /* Load required DLLs */
+    hKernel32 = GetModuleHandleW(L"kernel32.dll");
+    if (hKernel32)
     {
-        return;
+        pGetTickCount64 = (PFGETTICKCOUNT64)GetProcAddress(hKernel32, "GetTickCount64");
+        if (!pGetTickCount64)
+        {
+            hKernel32Vista = LoadLibraryW(L"kernel32_vista.dll");
+            if (hKernel32Vista)
+            {
+                pGetTickCount64 = (PFGETTICKCOUNT64)GetProcAddress(hKernel32Vista, "GetTickCount64");
+            }
+        }
     }
-    cSeconds = GetSeconds();
-    StringCchPrintfW(szBuf, _countof(szBuf), szStr,
-                     cSeconds / (60*60*24),
-                     (cSeconds / (60*60)) % 24,
-                     (cSeconds / 60) % 60,
-                     cSeconds % 60);
 
-    SetWindowTextW(hRosUptime, szBuf);
+    /* Show system uptime */
+    GetSystemUptime(hwndDlg);
+
+    /* Set update timer */
+    SetTimer(hwndDlg, ID_SYSUPTIME_UPDATE_TIMER, 1000 - (RosUptime % 1000), NULL);
```

Comment:
This call may be put in the GetSystemUptime() function (since said function takes the dialog handle as parameter as well); this would allow for *not* having `RosUptime` as an unnecessary global.

Reply:
`RosUptime` still needed to be global to get the correct update speed.